dev/buildtool/inspection_commands.py

```diff
@@ -12,100 +12,1154 @@
+                        server='bintray.api'),
+          'Failed on url=%s: %s' % (bintray_url, ex.message))
+    except Exception as ex:
+      raise
+    return headers, content
+
+  def list_bintray_packages(self, subject_repo):
+    path = 'repos/%s/packages' % subject_repo
+    base_url = 'https://api.bintray.com/' + path
+    result = []
     while True:
-      try:
-        urllib2.urlopen(url)
-        logging.debug('%s available', name_version)
-        return name_version
-      except urllib2.HTTPError as ex:
-        if not counted:
-          self.metrics.inc_counter(
-              'BintrayNotReady', {'repository': repository.name},
-              'Bintray not showing debian package that should be there.')
-          logging.warning('%s is not yet available on bintray: %s',
-                          name_version, ex)
-          counted = True
-
-        remaining = wait_until - time.time()
-        if remaining <= 0:
-          raise_and_log_error(
-              ResponseError(
-                  'Bintray not showing %s - giving up' % name_version,
-                  server='bintray'))
-        time.sleep(min(5, remaining))
-
-
-class WaitForDebiansFactory(RepositoryCommandFactory):
+      url = base_url + '?start_pos=%d' % len(result)
+      headers, content = self.fetch_bintray_url(url)
+      # logging.debug('Bintray responded with headers\n%s', headers)
+      total = headers.get('X-RangeLimit-Total', 0)
+      result.extend(['%s/%s' % (subject_repo, entry['name'])
+                     for entry in content])
+      if len(result) >= total:
+        break
+    return result
+
+  def query_bintray_package_versions(self, package_path):
+    path = 'packages/' + package_path
+    url = 'https://api.bintray.com/' + path
+    headers, content = self.fetch_bintray_url(url)
+    # logging.debug('Bintray responded with headers\n%s', headers)
+    package_name = package_path[package_path.rfind('/') + 1:]
+    return (package_name, content['versions'])
+
+  def difference(self, versions, target):
+    missing = []
+    for version in versions:
+      if not version in target:
+        missing.append(version)
+    return missing
+
+  def find_missing_jar_versions(self, jar_map, debian_map):
+    missing_jars = {}
+    prefix = 'spinnaker-'
+    for package, versions in debian_map.items():
+      if package.startswith(prefix):
+        key = package[len(prefix):]
+      if not key in jar_map:
+        key = package
+        if key == 'spinnaker-monitoring-daemon':
+          key = 'spinnaker-monitoring'
+        if not key in jar_map:
+          if key == 'spinnaker-monitoring-third-party':
+            continue
+          continue
+      missing = self.difference(versions, jar_map.get(key))
+      if missing:
+        missing_jars[key] = missing
+
+    return missing_jars
+
+  def find_missing_debian_versions(self, jar_map, debian_map):
+    missing_debians = {}
+    for package, versions in jar_map.items():
```

Comment:
Seems like we may be interested in debians listed in BOMs that are missing as well.

Reply:
This is just an artifact collection, independent of boms.
The two get tied together in the AuditArtifactVersions which this feeds.
The audit will feed all the boms into it. This command will produce an additional table saying "by the way, that debian was missing a  corresponding jar".